node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1454 @@
+					let message = ProvisionerMessage::ProvisionableData(
+						ProvisionableData::MisbehaviorReport(self.parent, report)
+					);
+
+					reports.push(message);
+				}
+			}
+		}
+
+		for report in reports.drain(..) {
+			self.send_to_provisioner(report).await?
+		}
+
+		Ok(())
+	}
+
+	async fn import_statement(&mut self, statement: SignedFullStatement) -> Result<(), Error> {
+		let idx = statement.validator_index() as usize;
+
+		if self.table_context.validators.len() > idx {
+			match statement.check_signature(
+				&self.table_context.signing_context,
+				&self.table_context.validators[idx],
+			) {
+				Ok(()) => {
+					let stmt = primitive_statement_to_table(&statement);
+
+					self.table.import_statement(&self.table_context, stmt);
+
+					self.issue_new_misbehaviors().await?;
+				}
+				Err(()) => {
+					return Err(Error::ValidationFailed(ValidationFailed));
+				}
+			}
+		};
+
+		Ok(())
+	}
+
+	async fn process_msg(&mut self, msg: CandidateBackingMessage) -> Result<(), Error> {
+		match msg {
+			CandidateBackingMessage::Second(_, candidate, pov) => {
+				// If the message is a `CandidateBackingMessage::Second`, sign and dispatch a
+				// Seconded statement only if we have not seconded any other candidate and
+				// have not signed a Valid statement for the requested candidate.
+				match self.seconded {
+					// This job has not seconded a candidate yet.
+					None => {
+						let candidate_hash = candidate.hash();
+
+						if !self.issued_validity.contains(&candidate_hash) {
+							if let Ok(ValidationResult::Valid) = self.validate_and_second(
+								candidate,
+								pov,
+							).await {
+								self.seconded = Some(candidate_hash);
+							}
+						}
+					}
+					// This job has already seconded a candidate.
+					Some(_) => {}
+				}
+			}
+			CandidateBackingMessage::Statement(_, statement) => {
+				self.maybe_validate_and_import(statement).await?;
+			}
+			CandidateBackingMessage::GetBackedCandidates(_, tx) => {
+				let backed = self.get_backed();
+
+				tx.send(backed).map_err(|_| oneshot::Canceled)?;
+			}
+		}
+
+		Ok(())
+	}
+
+	async fn maybe_validate_and_import(
+		&mut self,
+		statement: SignedFullStatement,
+	) -> Result<(), Error> {
+		let idx = statement.validator_index() as usize;
+
+		if self.table_context.validators.len() > idx {
+			if let Err(()) = statement.check_signature(
+				&self.table_context.signing_context,
+				&self.table_context.validators[idx],
+			) {
+				return Err(Error::InvalidSignature);
+			}
+		} else {
+			return Err(Error::InvalidSignature);
+		}
+
+		if let Statement::Seconded(candidate) = statement.payload() {
+			if candidate.parachain_index == self.assignment {
+				let pov = self.request_pov_from_distribution(candidate.clone().into()).await?;
+				let v = self.request_candidate_validation(candidate.clone(), pov).await?;
+
+				let statement = match v {
+					ValidationResult::Valid => Statement::Valid(candidate.hash()),
```

Comment:
As far as I can see `issued_validity` is not updated here as well.

It might be easier to just track all candidates we have issued statements for and what kind of statement we have issued.

The current implementation of the code causes us to attempt revalidation every time we see a new `Seconded` message, even if we have already validated that candidate or even seconded it ourselves. So just tracking all candidates we have already issued statement on (regardless of what type of statement that is) can already be used to prevent issuing conflicting statements.

Reply:
It also doesn't seem like the data is made available anywhere.